src/sysext/sysext.c

```diff
@@ -0,0 +1,987 @@
+                        log_info("No OS extensions found.");
+                        return 0;
+                }
+
+                t = table_new("name", "type", "path", "time");
+                if (!t)
+                        return log_oom();
+
+                HASHMAP_FOREACH(img, images) {
+                        r = table_add_many(
+                                        t,
+                                        TABLE_STRING, img->name,
+                                        TABLE_STRING, image_type_to_string(img->type),
+                                        TABLE_PATH, img->path,
+                                        TABLE_TIMESTAMP, img->mtime != 0 ? img->mtime : img->crtime);
+                        if (r < 0)
+                                return table_log_add_error(r);
+                }
+
+                (void) table_set_sort(t, (size_t) 0, (size_t) -1);
+
+                if (arg_json_format_flags & (JSON_FORMAT_OFF|JSON_FORMAT_PRETTY|JSON_FORMAT_PRETTY_AUTO))
+                        (void) pager_open(arg_pager_flags);
+
+                r = table_print_json(t, stdout, arg_json_format_flags);
+                if (r < 0)
+                        return table_log_print_error(r);
+
+                r = 0;
+                break;
+        }
+
+        case ACTION_MERGE: {
+                char **p;
+
+                /* In merge mode fail if things are already merged. (In --refresh mode below we'll unmerge if
+                 * we find things are already merged...) */
+                STRV_FOREACH(p, arg_hierarchies) {
+                        _cleanup_free_ char *resolved = NULL;
+
+                        r = chase_symlinks(*p, arg_root, CHASE_PREFIX_ROOT, &resolved, NULL);
+                        if (r == -ENOENT) {
+                                log_debug_errno(r, "Hierarchy '%s%s' does not exist, ignoring.", strempty(arg_root), *p);
+                                continue;
+                        }
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to resolve path to hierarchy '%s%s': %m", strempty(arg_root), *p);
+
+                        r = is_our_mount_point(resolved);
+                        if (r < 0)
+                                return r;
+                        if (r > 0)
+                                return log_error_errno(SYNTHETIC_ERRNO(EBUSY),
+                                                       "Hierarchy '%s' is already merged.", *p);
+                }
+
+                r = merge(images);
+                break;
+        }
+
+        case ACTION_REFRESH:
+                r = merge(images);
+                if (r < 0)
+                        return r;
+                if (r == 0) /* No images found? Then unmerge. The goal of --refresh is after all that after
+                             * having called it the merge status matches the installed extensions. */
+                        r = unmerge();
```

Comment:
I may misunderstand the concept of this tool, but is the order `merge` then `unmerge` correct? Why not `unmerge()` then `merge()`?

Reply:
so the merge() actually does an unmerge internally, once it managed to dissect/mount/prepare the new overlayfs. In which case merge() returns > 0. But if there are no extensions installed whatsoever, it returns == 0 and doesn't do a thing. For the --refresh to always leave things in a state matching the current set of installed extensions we thus need to call unmerge() in this case, so that whatever was mounted before gets unmounted.

i.e. the merge() covers the cases where there were no extensions before and now are and where there were some extensions before and still are. But the cases where there are no longer any extensions needs to be covered too, and that's why we call unmerge() after the merge() if we detect this condition.

i.e.:

currently /usr is not overlayfs merged + extensions installed → merge(), no unmerge() whatsoever
currently /usr is overlayfs merged + extensions installed → merge() with implicit unmerge()
currently /usr is overlayfs merged + no extensions installed anymore → unmerge()
currently /usr is not overlayfs merged + no extensions installed → NOP

The comment above the unmerge() was supposed to explain this, but I figure it's not good enough at that?

Reply:
Ah, thank you for the detailed explanation. I'd like to also have the summary in the code comment.